libmamba/include/mamba/util/iterator.hpp

```diff
@@ -0,0 +1,400 @@
+        using base_type = select_filter_iterator_base_t<Predicate, Iterator>;
+
+        using reference = typename base_type::reference;
+        using pointer = typename base_type::pointer;
+        using difference_type = typename base_type::difference_type;
+        using iterator_category = typename base_type::iterator_category;
+
+        filter_iterator() = default;
+
+        filter_iterator(Iterator iter, Iterator begin, Iterator end)
+            : m_pred()
+            , m_iter(iter)
+            , m_begin_limit(--begin)
+            , m_end(end)
+        {
+            satisfy_predicate();
+        }
+
+        template <class Pred>
+        filter_iterator(Pred&& pred, Iterator iter, Iterator begin, Iterator end)
+            : m_pred(std::forward<Pred>(pred))
+            , m_iter(iter)
+            , m_begin_limit(--begin)
+            , m_end(end)
+        {
+            satisfy_predicate();
+        }
+
+        self_type& operator++()
+        {
+            ++m_iter;
+            satisfy_predicate();
+            return *this;
+        }
+
+        self_type& operator--()
+        {
+            --m_iter;
+            while (m_iter != m_begin_limit && !m_pred(*m_iter))
+            {
+                --m_iter;
+            }
+            return *this;
+        }
+
+        template <class It = Iterator>
+        enable_random_access_iterator<It, self_type>
+        operator+=(difference_type n)
+        {
+            if (n >= 0)
+            {
+                advance(n);
+            }
+            else
+            {
+                step_back(n);
+            }
+            return *this;
+        }
+
+        template <class It = Iterator>
+        enable_random_access_iterator<It, self_type>
+        operator-=(difference_type n)
+        {
+            if (n >= 0)
+            {
+                step_back(n);
+            }
+            else
+            {
+                advance(n);
+            }
+            return *this;
+        }
+
+        template <class It = Iterator>
+        enable_random_access_iterator<It, difference_type>
+        operator-(const It& rhs) const
+        {
+            It tmp = rhs;
+            difference_type res = 0;
+            while (tmp++ != *this)
+            {
+                ++res;
+            }
+            return res;
+        }
+
+        reference operator*() const
+        {
+            return *m_iter;
+        }
+
+        pointer operator->() const
+        {
+            return m_iter.operator->();
+        }
+
+        bool equal(const self_type& rhs) const
+        {
+            return m_pred == rhs.m_pred && m_iter == rhs.m_iter;
+        }
+
+        template <class It = Iterator>
+        enable_random_access_iterator<It, bool>
+        less_than(const self_type& rhs) const
+        {
+            return m_pred == rhs.m_pred && m_iter < rhs.m_iter;
+        }
+
+    private:
+
+        void advance(difference_type n)
```

Comment:
Why not use `std::advance` instead?
also, it's the same code with both directions so I would gather step_back and advance together, the only difference is the direction, which is already given by `n`

Reply:
Because `std::advance` implementation calls `operator+=` for random access iterators, leading tooooo:
Stack bomb, stack bomb, yoooooou're my stack bomb:D!